Fix #1507: Add new Teams: Add or update team repository endpoint

andygrunwald opened the pull request:
See https://github.com/PyGithub/PyGithub/issues/1507

Side Note:
This PR is untested for now because I don't know how to use a local python library (yet). However, I want to get feedback on this PR if this is the way if this is acceptable.

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -234,13 +243,23 @@ def get_repo_permission(self, repo):
         except UnknownObjectException:
             return None
 
+    @deprecated(
```
This is missing an import: from deprecated import deprecated at the top of a file.

andygrunwald replied:
Thanks. Fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/0373fcde52f3b71da75e3eb277c52b265a315793

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -234,13 +243,23 @@ def get_repo_permission(self, repo):
         """
-        :calls: `PUT /teams/:id/repos/:org/:repo <http://developer.github.com/v3/orgs/teams>`_
+        This API call is deprecated. Use `update_team_repository` instead:
+        https://developer.github.com/v3/teams/#add-or-update-team-repository-legacy
```
I think we can ignoring changing the doc string --- or the method to add a blank line, and just add the deprecation message.

andygrunwald replied:
Thanks. Fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/8896c88e1dd4b6a7e6ef01b51d1223f21e5362a8

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -249,6 +268,26 @@ def set_repo_permission(self, repo, permission):
+        # Original note from the docs:
+        # Note that, if you choose not to pass any parameters, you'll need to set Content-Length to zero when calling out to this endpoint. For more information, see "HTTP verbs."
+
+        assert isinstance(repo, github.Repository.Repository), repo
```
Why would we call this with no parameters? I don't understand.

Further, you could make this API a little simpler to use to by also accepting the repo name as a string.

andygrunwald replied:
> Why would we call this with no parameters? I don't understand.

This sentence is a copy of the docs. See https://developer.github.com/v3/teams/#add-or-update-team-repository
In the docs, only the permission parameter is an optional string. If no permission is added, the API will behave like:

> If no permission is specified, the team's permission attribute will be used to determine what permission to grant the team on this repository.

Do you prefer owner + repo in one string (e.g. `PyGithub/PyGithub`) or owner and repo in two parameters (Owner: `PyGithub`, repo: `PyGithub`) or only the repo, because the API assumes that the owner == org?

s-t-e-v-e-n-k replied:
I wasn't going to get so down into the weeds as that, I was just suggesting you change the assert to accept a str as well and passing that in.

andygrunwald replied:
Added in https://github.com/PyGithub/PyGithub/pull/1509/commits/4b7126cbb6d991967b50ca5ad86e500bc6798e00

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -249,6 +268,26 @@ def set_repo_permission(self, repo, permission):
+        headers, data = self._requester.requestJsonAndCheck(
+            "PUT", self.organization_url + "/teams/" + str(self.id)  + "/repos/" + repo._identity, input=put_parameters
+        )
+
```
Do we want to return a new object here? I guess it's github.Permission.Permission.

andygrunwald replied:
The official API returns only a `Status: 204 No Content`.
Maybe it would be good to return a bool if this was successful or not? What doy ou think?

s-t-e-v-e-n-k replied:
Ah, good point. Have a look at Repository.py to get a hint about we use that pattern.

andygrunwald replied:
Fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/1e728aa738d97e075d8f0347e25931569e62fc6c

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -447,6 +487,8 @@ def _useAttributes(self, attributes):
         if "description" in attributes:  # pragma no branch
             self._description = self._makeStringAttribute(attributes["description"])
+        if "organization" in attributes and "url" in attributes["organization"]:  # pragma no branch
+            self._organization_url = self._makeStringAttribute(attributes["organization"]["url"])
```
Oh, is it an actual organization? We should just set that then --

```
self._organization = self._makeClassAttribute(github.Organization.Organization, attributes["organization"])

s-t-e-v-e-n-k replied:
We already set _organization in fact! Why are you adding a _url parameter?

andygrunwald replied:
Oh. Indeed. Thanks for pointing this out. Fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/b7bea98e05a4fdc7d1de46d2d2e3b1697a0ea7e5

s-t-e-v-e-n-k commented:
Er, Fixes #1491 indeed...

andygrunwald commented:
Thanks @s-t-e-v-e-n-k for the review. I have made some changes and asked a few questions.

> Also, the commit body should include "Fixes #1507" not the commit title.

Can you squash the commit once it is good to merge? Otherwise, I would have to make a push force.

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -234,13 +235,20 @@ def get_repo_permission(self, repo):
         :param permission: string
         :rtype: None
         """
+
```
This blank line you've added can also be removed.

andygrunwald replied:
Fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/f61df4a51357d89c2baa1b074a31294535b3a241

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -234,13 +235,20 @@ def get_repo_permission(self, repo):
+    @deprecated(
+        reason="""
+        Team.set_repo_permission() is deprecated, use
+        Team.update_team_repository() instead.
```
I wonder if this fits on one line. It's no issue if it does not.

andygrunwald replied:
Done in https://github.com/PyGithub/PyGithub/pull/1509/commits/076e25776ce84f4dc96f7be9aba5ee3e120527af

s-t-e-v-e-n-k commented:
Further, you've added some linting failures, which is why the 3.6 job fails, have a look at the ouput of tox -e lint

andygrunwald commented:
@s-t-e-v-e-n-k Linting errors fixed in https://github.com/PyGithub/PyGithub/pull/1509/commits/6c6cd17bcbcc48bd8ec96c5f025f8727e2210136

And this completes the second round of iteration. Would you mind checking the code again?

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -249,6 +257,37 @@ def set_repo_permission(self, repo, permission):
+
+        assert repo is isinstance(repo, github.Repository.Repository) or isinstance(
+            repo, str
+        ), repo
```
Also assert permission is a string. In which case, the comment is also not needed, and can be removed too.

andygrunwald replied:
Adjusted in https://github.com/PyGithub/PyGithub/pull/1509/commits/32db684d6cce40eca8ba0dc62be4774a187dfcd2

s-t-e-v-e-n-k commented on github/Team.py:
```diff
@@ -249,6 +257,37 @@ def set_repo_permission(self, repo, permission):
+            repo, str
+        ), repo
+        repo_url_param = repo
+        if repo is github.Repository.Repository:
```
``if isinstance(repo, github.Repository.Repository):``

andygrunwald replied:
Thanks, changed in https://github.com/PyGithub/PyGithub/pull/1509/commits/e51fa257bd254e506cfd6c1f93219c8f802f84ba

andygrunwald commented:
@s-t-e-v-e-n-k I think we are moving in the right direction.
With the unit test, I fixed two small bugs (assertion + API url).

I recorded a unit test, however, it seems to fail. Here is what i did:
- Added credentials into the `GithubCredentials.py` file
- Changed the `setUp` to use an organization and team of mine
- Executed `pytest tests/Team.py -k testUpdateTeamRepository --record --auth_with_token`

Any guidance on how to record a test that it will succeed the CI checks?

s-t-e-v-e-n-k commented:
It's a sharp edge with the replaying, I'm afraid -- if you look at the text file that the recording mode wrote out, you'll see the hostname of the request, headers sent and the reply, etc -- the problem is your replay data has token auth in the headers sent, and other tests mock password auth. It should be an obvious fix by looking at another replay data file.

s-t-e-v-e-n-k commented on tests/Team.py:
```diff
@@ -110,6 +110,11 @@ def testRepoPermission(self):
+    def testUpdateTeamRepository(self):
+        repo = self.org.get_repo("test-repo")
+        res = self.team.update_team_repository(repo, "admin")
+        self.assertTrue(res)
```
Just ``self.assertTrue(self.team.update_team_repository(repo, "admin"))`` it's readable enough as one line.

s-t-e-v-e-n-k commented:
@andygrunwald Do you have time to get back to this? It's so close!

andygrunwald commented:
@s-t-e-v-e-n-k I still have it on my list to finish it. Sorry for taking so long. Private topics are filling up my time right now.

LenoxGOLD commented:
✌️

andygrunwald commented:
@s-t-e-v-e-n-k I have adjusted the replay payload manually. However, the error messages don't make sense to me, Can you guide me a bit here to make the test pass?

andygrunwald commented:
@s-t-e-v-e-n-k Any news on this?

s-t-e-v-e-n-k commented:
@andygrunwald I'm so sorry for forgetting to look at this! ``TypeError: can only concatenate str (not "NoneType") to str`` looks like the proximate cause when constructing the URL.

andygrunwald commented:
@s-t-e-v-e-n-k I don't understand on how to fix this error. It seems to be related to the replay data file.
Not sure what is missing or what is wrong. The replay file seems quite similar to the others.
Would you mind having a look at this?

How to execute this single unit test:
```
$ pytest tests/Team.py -k testUpdateTeamRepository
```

Output right now:
```
================================================================================================================== ERRORS ==================================================================================================================
____________________________________________________________________________________________ ERROR at teardown of Team.testUpdateTeamRepository ____________________________________________________________________________________________
tests/Framework.py:297: in tearDown
    self.__closeReplayFileIfNeeded()
tests/Framework.py:325: in __closeReplayFileIfNeeded
    self.assertEqual(readLine(self.__file), "")
E   AssertionError: 'https' != ''
E   - https
E   +
================================================================================================================= FAILURES =================================================================================================================
______________________________________________________________________________________________________ Team.testUpdateTeamRepository _______________________________________________________________________________________________________

self = <tests.Team.Team testMethod=testUpdateTeamRepository>

    def testUpdateTeamRepository(self):
        repo = self.org.get_repo("FatherBeaver")
>       self.assertTrue(self.team.update_team_repository(repo, "admin"))

tests/Team.py:115:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Team(name="Team created by PyGithub", id=189850), repo = Repository(full_name="BeaverSoftware/FatherBeaver"), permission = 'admin'

    def update_team_repository(self, repo, permission):
        """
        :calls: `PUT /orgs/:org/teams/:team_slug/repos/:owner/:repo <https://developer.github.com/v3/teams/#add-or-update-team-repository>`_
        :param repo: string or :class:`github.Repository.Repository`
        :param permission: string
        :rtype: bool
        """
        assert isinstance(repo, github.Repository.Repository) or isinstance(
            repo, str
        ), repo
        assert isinstance(permission, str), permission
        repo_url_param = repo
        if isinstance(repo, github.Repository.Repository):
            repo_url_param = repo._identity
        put_parameters = {
            "permission": permission,
        }
        status, _, _ = self._requester.requestJson(
            "PUT",
            self.organization.url
            + "/teams/"
            + self.slug
            + "/repos/"
>           + repo_url_param,
            input=put_parameters,
        )
E       TypeError: can only concatenate str (not "NoneType") to str

github/Team.py:283: TypeError
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/Team.py::Team::testUpdateTeamRepository - TypeError: can only concatenate str (not "NoneType") to str
ERROR tests/Team.py::Team::testUpdateTeamRepository - AssertionError: 'https' != ''
================================================================================================= 1 failed, 9 deselected, 1 error in 0.24s =================================================================================================
```

s-t-e-v-e-n-k commented:
The error at the top is a side effect of the test failure, the fixtures are a little brittle, so they tend to explode when something unexpected happens.

So the proximate cause is either self.slug, repo_url_param is None. ``repo_name = repo._identity if isintance(github.Repository.Repository) else repo`` is how that line is expressed elsewhere.

I would suggest either using pdb to step and see what's set or not, or adding two lines to the method to see which goes bang:
```
assert self.slug is not None
assert repo_url_param is not None
``

andygrunwald commented:
@s-t-e-v-e-n-k Thanks for the hint.
`self.slug` was not set, because it was not part of the test stub. That's why it worked in production, but not in the unit tests.
In the last couple of commits, I
- set the correct API URL link
- fixed a formatting issue of python 3.6
- set team.slug in the unit test stub data

Would you mind to review it again?

pascal-hofmann reviewed (commented):
@s-t-e-v-e-n-k This looks good to me now. Can I approve and merge or should I wait for your re-review?
